### PR TITLE
feat: celestia-da client for rollup txs against data availability root

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

NOTE: This PR is a work-in-progress and contains now only boilerplate code. This is a starting point and has not been tested yet! 

Initial proof-of-concept effort for celestia data availability client. The idea to is reuse as much 07-tendermint code as possible, changing only the parts that matter, i.e. consensus state data and proof verification against the data availability root.

The celestia-core `ShareProof` is the type we are interested in for proof verification, verified from this gist: https://gist.github.com/damiannolan/b83ecbaad2c921454e8ae42906816eb5

How does this diverge from 07-tendermint?

1. `UpdateState` stores `ConsensusState` data with data availability root - `header.GetDataHash()` instead of `header.GetAppHash()`. 
2. `VerifyMembership` unmarshals proof bytes to `ShareProof` to validate against data hash.

Right now I just named this 07-celestia due to its similarities. We can use a different num prefix tho - feel free to suggest.

ref: #5318 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/light-clients/07-celestia/client_state.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	return cs.BaseClient.Initialize(ctx, cdc, clientStore, consensusState)
+}
+
+// Validate implements exported.ClientState.
+func (cs *ClientState) Validate() error {
+	return cs.BaseClient.Validate()
+}
+
+// VerifyMembership implements exported.ClientState.
+// TODO: Revise and look into delay periods for this.
```
Comment (damiannolan) [inline modules/light-clients/07-celestia/client_state.go:49]: I think we can add a `keeper` pkg and control delay period values with gov gated on-chain params.

Comment (damiannolan) [reply modules/light-clients/07-celestia/client_state.go:49]: https://github.com/cosmos/ibc-go/issues/5852

**Review by AdityaSripal — COMMENTED** [review]
LGTM, the proof code should probably be reviewed by @nashqueue

**Comment by damiannolan** [comment]
> LGTM, the proof code should probably be reviewed by @nashqueue

Yep. I copied share proof protos and implementation temporarily but I believe they should be made available in this repo at some point https://github.com/celestiaorg/go-square. Would like to follow up with some unit tests also of course. 

This PR is now targeting a feat branch. I think it should be possible to also make this client a standalone go module as well 👍🏻

**Comment by damiannolan** [comment]
Marking as r4r since target is feat branch. We can follow up with more testing PRs etc.

**Review by AdityaSripal — APPROVED** [review]
Approving as no changes made since last review where the code looked ready to go for further steps in rollkit integration

**Comment by damiannolan** [comment]
> One neat refactor would be to import https://github.com/celestiaorg/go-square instead of reimplementing the proofs and formats.

Agree, would be nice to consume a lib. Can you point me to where the `ShareProof` code is in `go-square`, I can't find it - maybe its not available on main? Or is there some other proof construction with parity to `ShareProof` that should be used?

File: modules/light-clients/07-celestia/client_state.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		)
+	}
+
+	if err := verifyDelayPeriodPassed(ctx, clientStore, height, delayTimePeriod, delayBlockPeriod); err != nil {
+		return err
+	}
+
+	var shareProofProto ShareProof
+	if err := cdc.Unmarshal(proof, &shareProofProto); err != nil {
+		return err
```
Comment (crodriguezvega) [inline modules/light-clients/07-celestia/client_state.go:65]:
```suggestion
                return errorsmod.Wrapf(commitmenttypes.ErrInvalidProof, "could not unmarshal share proof: %v", err)
```

**Comment by damiannolan** [comment]
> Nice work, @damiannolan. Great idea to reuse 07-tendermint's logic.
> 
> Happy to merge this to the feat branch and continue on followup PRs.
> 
> Should we add tests for `UpdateState` and `VerifyMembership`?

Happy to follow up on comments with smaller PRs post-merge. 
Retargetting to `feat/celestia-da-client` branch and merging 👍🏻

**Pull request merged**